### PR TITLE
168 Support management of certificates in aem global trust store

### PR DESCRIPTION
**Pull request by jan-kowalczyk-wttech**

### CODE REVIEW

File: pkg/gts/gts_certificate.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+		return false, err
+	}
+
+	notAfter, err := c.GetNotAfter()
+
+	if err != nil {
+		return false, err
+	}
+
+	fmt.Println(c.Subject)
```
Comment (krystian-panek-wttech) [inline pkg/gts/gts_certificate.go:51]: Tmp code i guess 😁

File: pkg/gts/gts_status.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+		if s.Certificates[i].Alias == alias {
+			return &s.Certificates[i]
+		}
+	}
+
+	return nil
+}
+
+func UnmarshalStatus(readCloser io.ReadCloser) (*Status, error) {
+
```
Comment (krystian-panek-wttech) [inline pkg/gts/gts_status.go:44]: Format of data should never change! This is a very often cause of app errors. Format of data is a contract between apps.

File: pkg/instance.go [hunk 3]
```diff
@@ -21,16 +21,17 @@ type Instance struct {
...
-	workflowManager *WorkflowManager
+	local            *LocalInstance
+	http             *HTTP
+	status           *Status
+	repo             *Repo
+	osgi             *OSGi
+	sling            *Sling
+	crypto           *Crypto
+	ssl              *SSL
+	globalTrustStore *GTSManager
```
Comment (krystian-panek-wttech) [inline pkg/instance.go:32]: gtsStore

File: pkg/instance_manager.go [hunk 4]
```diff
@@ -251,6 +251,7 @@ func (im *InstanceManager) New(id, url, user, password string) *Instance {
 	res.sling = NewSling(res)
 	res.crypto = NewCrypto(res)
 	res.ssl = NewSSL(res)
+	res.globalTrustStore = NewGlobalTrustStoreMananger(res)
```
Comment (krystian-panek-wttech) [inline pkg/instance_manager.go:254]: gts & NewGTS()

File: cmd/aem/gts.go [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+
+	cmd.Flags().String("password", "", "Password to Global Trust Store")
+	_ = cmd.MarkFlagRequired("password")
+
+	return cmd
+}
+
+func (c *CLI) gtsCertCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "certificate",
```
Comment (krystian-panek-wttech) [inline cmd/aem/gts.go:82]: how about Aliases `cert` ?

File: cmd/aem/gts.go [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+	cmd.Flags().String("path", "", "Certificate file path (PEM|DER format)")
+	_ = cmd.MarkFlagRequired("path")
+
+	return cmd
+}
+
+func (c *CLI) gtsCertRemoveCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "remove",
+		Aliases: []string{},
```
Comment (krystian-panek-wttech) [inline cmd/aem/gts.go:132]: please review existing aliases for other `remove` commands and fulfill them here

File: cmd/aem/gts.go [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+
+	cmd.Flags().String("alias", "", "Certificate Alias")
+	_ = cmd.MarkFlagRequired("alias")
+
+	return cmd
+}
+
+func (c *CLI) gtsCertReadCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "read",
```
Comment (krystian-panek-wttech) [inline cmd/aem/gts.go:165]: add Aliases which are already set for other `read`-like commands

File: cmd/aem/gts.go [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+		Aliases: []string{},
+		Short:   "Read certificate from Global Trust Store",
+		Run: func(cmd *cobra.Command, args []string) {
+			instance, err := c.aem.InstanceManager().One()
+			if err != nil {
+				c.Error(err)
+				return
+			}
+
+			certifiacteAlias, _ := cmd.Flags().GetString("alias")
```
Comment (krystian-panek-wttech) [inline cmd/aem/gts.go:175]: you are in the scope of certificate (method name has certificate keyword)  so your variable could be just `alias`

File: pkg/common/cert/cert.go [hunk 9]
```diff
@@ -0,0 +1,35 @@
+package cert
+
+import (
+	"encoding/pem"
+	"fmt"
+	"os"
+)
+
+func CreateTmpDerFileBasedOnPem(block *pem.Block) (*string, func(), error) {
```
Comment (krystian-panek-wttech) [inline pkg/common/cert/cert.go:9]: candidate for `common.certx` or sth ;)

File: pkg/gts/gts_certificate.go [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+type Certificate struct {
+	Alias        string      `json:"alias"`
+	Subject      string      `json:"subject"`
+	Issuer       string      `json:"issuer"`
+	NotBefore    string      `json:"notBefore"`
+	NotAfter     string      `json:"notAfter"`
+	SerialNumber json.Number `json:"serialNumber"`
+}
+
+func (c *Certificate) GetNotBefore() (time.Time, error) {
```
Comment (krystian-panek-wttech) [inline pkg/gts/gts_certificate.go:25]: in go getter are without `Get` .. just `NotBefore()`

File: pkg/common/cert/cert.go [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+	tempDerFile, err := os.CreateTemp("", "tmp-cert-*.der")
```
Comment (absa-piotr-andruszkiewicz) [inline pkg/common/cert/cert.go:10]: Where do you clear/delete the temp file? In my solution it got auto-deleted thanks to `defer` keyword

Comment (jan-kowalczyk-wttech) [reply pkg/common/cert/cert.go:10]: this function to defer is returned, and it must be defered by function that run "CreateTmpDerFileBasedOnPem". I didn't found any other possibilities to remove duplication of code

File: README.md [hunk 12]
```diff
@@ -544,6 +545,37 @@ See the reference documentation: [AEM 6.5 > Administering Guide > SSL by Default
...
+AEM Compose supports managing the trust store of AEM instances.
+
+This feature supports:
+- creation of the general trust store if it does not exist
+```shell
+sh aemw gts create --password PASSWORD_HERE --instance-author
+```
+- getting the general trust store status
+```shell
+sh aemw gts status --instance-author
```
Comment (krystian-panek-wttech) [inline README.md:559]: I prefer `-A`

File: README.md [hunk 13]
```diff
@@ -544,6 +545,37 @@ See the reference documentation: [AEM 6.5 > Administering Guide > SSL by Default
...
+```
+
+- adding a certificate to the general trust store
+```shell
+sh aemw gts certificate add --path PATH_TO_CERTIFICATE --instance-author
```
Comment (krystian-panek-wttech) [inline README.md:564]: just `<path>` maybe?

File: README.md [hunk 14]
```diff
@@ -544,6 +545,37 @@ See the reference documentation: [AEM 6.5 > Administering Guide > SSL by Default
...
+```
+This command will add a certificate to the general trust store only if not exists in trust store and will return the alias of the certificate.
+Command `certificate add` supports certificates in PEM and DER formats.
+
+- reading a certificate from the general trust store (by alias)
+```shell
+sh aemw gts certificate read --alias CERT_ALIAS --instance-author
```
Comment (krystian-panek-wttech) [inline README.md:571]: CERT_ALIAS = `<alias>` or `<string>`

File: cmd/aem/gts.go [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+		Short:   "Get status of Global Trust Store",
+		Aliases: []string{"show", "get", "read", "describe", "ls"},
+		Run: func(cmd *cobra.Command, args []string) {
+			instance, err := c.aem.InstanceManager().One()
+			if err != nil {
+				c.Error(err)
+				return
+			}
+
+			result, err := instance.GlobalTrustStore().Status()
```
Comment (krystian-panek-wttech) [inline cmd/aem/gts.go:30]: `instance.GTS()`

File: cmd/aem/gts.go [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+
+			if err != nil {
+				c.Error(err)
+				return
+			}
+
+			if result.Created == true {
```
Comment (krystian-panek-wttech) [inline cmd/aem/gts.go:37]: remove `== true` 

is the code linted?

File: cmd/aem/gts.go [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+	cmd.Flags().String("password", "", "Password to Global Trust Store")
+	_ = cmd.MarkFlagRequired("password")
+
+	return cmd
+}
+
+func (c *CLI) gtsCertCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "certificate",
+		Aliases: []string{"certx", "crt"},
```
Comment (krystian-panek-wttech) [inline cmd/aem/gts.go:84]: cert, crt

File: pkg/gts_manager.go [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+	"os"
+)
+
+type GTSManager struct {
+	instance *Instance
+}
+
+func NewGTSMananger(instance *Instance) *GTSManager {
+	return &GTSManager{
+		instance: instance,
```
Comment (krystian-panek-wttech) [inline pkg/gts_manager.go:19]: you could inline these 3 into 1

File: pkg/gts_manager.go [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+const (
+	GTSPathJson = "/libs/granite/security/truststore.json"
+	GTSPath     = "/libs/granite/security/post/truststore"
+)
+
+func (gtsManager *GTSManager) Status() (*gts.Status, error) {
+	response, err := gtsManager.instance.http.Request().Get(GTSPathJson)
+
+	if err != nil {
+		return nil, fmt.Errorf("%s > cannot read global trust store: %w", gtsManager.instance.ID(), err)
```
Comment (krystian-panek-wttech) [inline pkg/gts_manager.go:32]: in other places you have capitalized "Global Trust Store" ; be consistent

File: pkg/gts_manager.go [hunk 20]
```diff
@@ -0,0 +1,181 @@
...
+
+	if err != nil {
+		return nil, false, fmt.Errorf("%s > failed to read certificate file '%s': %w", gtsManager.instance.ID(), certificateFilePath, err)
+	}
+
+	pemBlock, _ := pem.Decode(cretificateData)
+	if pemBlock != nil {
+		tmpDerFileNameBasedOnPemPath, functionToDefer, err := certx.CreateTmpDerFileBasedOnPem(pemBlock)
+
+		defer functionToDefer()
```
Comment (krystian-panek-wttech) [inline pkg/gts_manager.go:88]: cleanCallback ?

**Review by krystian-panek-wttech — APPROVED** [review]

**Pull request merged**